spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,800 @@
+    private final Map<XField, XMethod> declaredLockObjects = new HashMap<>();
+    private final Map<XField, XMethod> inheritedLockObjects = new HashMap<>();
+    private final MultiMap<XField, XMethod> declaredLockAccessors = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> lockAccessorsInHierarchy = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> lockUsingMethodsInHierarchy = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> declaredInheritedFieldAccessingMethods = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> potentialObjectBugContainingMethods = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> potentialInheritedBugContainingMethods = new MultiMap<>(HashSet.class);
+    // Backing collections
+    private final Set<XField> badBackingCollections = new HashSet<>();
+    private final Set<XField> inheritableBackingCollections = new HashSet<>();
+    private final MultiMap<XField, BugInfo> declaredCollectionLockObjects = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, BugInfo> inheritedCollectionLockObjects = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> declaredCollectionAccessors = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XMethod> collectionAccessorsInHierarchy = new MultiMap<>(HashSet.class);
+    private final MultiMap<XField, XField> backingCollections = new MultiMap<>(HashSet.class);
+
+    public FindImproperSynchronization(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    private static boolean isWrapperImplementation(MethodDescriptor methodDescriptor) {
+        return WRAPPER_IMPLEMENTATIONS.containsKey(methodDescriptor);
+    }
+
+    private static Integer getStackOffset(MethodDescriptor methodDescriptor) {
+        return WRAPPER_IMPLEMENTATIONS.get(methodDescriptor);
+    }
+
+    private static boolean isImmutableReturner(String signature) {
+        return IMMUTABLE_RETURNERS.contains(signature);
+    }
+
+    @Override
+    public void visit(Method obj) {
+        try {
+            analyzeAssignments(obj, getClassContext());
+        } catch (CFGBuilderException e) {
+            AnalysisContext.logError("Error building CFG for " + obj, e);
+        }
+
+        if (isInitializerMethod(obj.getName())) {
+            return;
+        }
+
+        XMethod xMethod = getXMethod();
+        if (xMethod.isPublic() && xMethod.isSynchronized()) {
+            if (xMethod.isStatic()) {
+                bugReporter.reportBug(new BugInstance(
+                        this, "US_UNSAFE_STATIC_METHOD_SYNCHRONIZATION", NORMAL_PRIORITY)
+                        .addClassAndMethod(this));
+            }
+            synchronizedMethods.add(xMethod);
+        }
+
+        SignatureParser signature = new SignatureParser(xMethod.getSignature());
```

Comment:
The `signature` field is only used first in line 205. Can you please move this field closer to the first usage of it?

Reply:
Yes, definitively.